### PR TITLE
Xy block

### PR DESCRIPTION
**Pull request by dmbates**
- Replace the last two blocks, `X` and `y`, in the terms field with a single `femat` matrix that includes both `X` and `y`
- This cuts down on the number of calls to multiplication, etc. in `updateL!` and allows multi-threaded BLAS to be more effective
- This PR is incomplete

### CODE REVIEW

File: src/linearmixedmodel.jl
Comment (palday) [inline src/linearmixedmodel.jl:739]: There isn't a field `fe`, which ties into another thing I was thinking. Namely, that we should think about making the documentation for the `femat` vs `feterm` field very clear (and also clarify why we need both).

Comment (dmbates) [reply src/linearmixedmodel.jl:739]: Yes, absolutely.  I would still keep open the possibility of not using this change.  I think it will be helpful but it may cause too much pain with the sparse model matrices.  We'll see.  I just wanted to get it to a place where you can look at it.

**Comment by dmbates** [comment]
- Changes in the GLMM code to accommodate the new structures.
- Several changes are because the GLM package specializes some methods to `Vector{T}`.  If those were changed to `AbstractVector{T}` the changes would not be necessary.
- Some of the changes of arguments like `foo.bar.baz'` to `adjoint(foo.bar.baz)` were simply to get the linter to shut up.  They were being flagged as syntax errors despite the fact that they were not syntax errors.

**Comment by palday** [comment]
Thanks for the explanatory comments. :) Should we open a PR against GLM for changing loosening the type requirements?

File: test/pls.jl
Comment (palday) [inline test/pls.jl:95]: thanks for adding the comment with the change!

**Comment by dmbates** [comment]
@palday I am done tinkering.  Now if you could just do the heavy lifting of resolving conflicts, etc.

File: src/linalg.jl
Comment (palday) [inline src/linalg.jl:34]: was this just because of the linter complaining?

Comment (dmbates) [reply src/linalg.jl:34]: Yes.  I think they may have even fixed the issue in the linter used with VS Code now.  Your call on whether you want to revert this change or not.

File: src/Xymat.jl
Comment (palday) [inline src/Xymat.jl:6]:
```suggestion
Typically, an `FeTerm` represents the design matrix for the fixed effects. 

!!! note
    `FeTerm` is not the same as [`FeMat`](@ref)!

# Fields
```

I don't know if `FeTerm` is the best name because it doesn't work like a `StatsModels.AbstractTerm`. Maybe make this `FeMat` and `FeMat` into `FeResponseMat` or `XYMat` or `XyMat`?

File: src/Xymat.jl
Comment (palday) [inline src/Xymat.jl:82]:
```suggestion

Typically, an `FeMat` represents the fixed-effects design matrix with the response (`y`) concatenated as a final column.

!!! note
    `FeMat` is not the same as [`FeTerm`](@ref).

# Fields
```

File: src/Xymat.jl
Comment (palday) [reply src/Xymat.jl:6]: Having now also talked to @kleinschmidt , I think we should change at least one of the names. 

More generally, I'm not sure `FeTerm` and `ReTerm` are ideal, but we're internally consistent in our use of those even though they're not `Term`s in the StatsModel sense, but that latter bit is for a different PR. We could change those names or make them more term like, but again, that's something for a later day.

File: src/Xymat.jl
Comment (palday) [inline src/Xymat.jl:75]: The pivot vector will always be the same length as the dimension, so this reduces to checking the rank against the dimension.

File: src/Xymat.jl
Comment (palday) [inline src/Xymat.jl:104]:
```suggestion
Base.length(A::FeMat) = length(A.xy)
```

More consistent if weight-independent ops always operate on the nonweighted field.

File: src/Xymat.jl
Comment (palday) [inline src/Xymat.jl:130]: We have a few of these `Base` methods we're extending just by delegating -- should we put them in a loop and delegate on the symbols? 

No strong preference either way on my end.

File: src/Xymat.jl
Comment (palday) [inline src/Xymat.jl:84]:
```suggestion
- `wtxy`: (possibly) weighted copy of `xy` (shares storage with `xy` until weights are applied)
```

File: src/linearmixedmodel.jl
Comment (palday) [inline src/linearmixedmodel.jl:11]:
```suggestion
* `feterm`: the fixed-effects model matrix as an `FeTerm{T}`
```

File: src/linearmixedmodel.jl
Comment (palday) [inline src/linearmixedmodel.jl:298]:
```suggestion
end

function createAL(reterms::Vector{AbstractReMat{T}}, Xy::FeMat{T}) where {T}
```

File: src/linearmixedmodel.jl
Comment (palday) [inline src/linearmixedmodel.jl:296]: Since all the calls to `pushALblock!` have the `blk` wrapped in `densify`, I wonder if it might make sense to have a `copy` option in that instead of `deepcopy` here. For matrices that are made dense, we're already dealing with a deep copy and so we add an extra copy here. For non dense matrices, we still need the copy of course.

File: src/linearmixedmodel.jl
Comment (palday) [inline src/linearmixedmodel.jl:311]: I'm very curious whether we any sparse FE design matrix actually gets stored as sparse after the product here....

File: src/linearmixedmodel.jl
Comment (palday) [inline src/linearmixedmodel.jl:356]: more general question (i.e. not a change for this PR): I wonder if we should better document when our accessor methods are returning views / shallow copies and when they're returning deep copies.

File: src/linearmixedmodel.jl
Comment (palday) [inline src/linearmixedmodel.jl:9]: with the new Xy storage, `allterms` no longer provides a speed boost?

File: src/linearmixedmodel.jl
Comment (palday) [inline src/linearmixedmodel.jl:593]: I've lost the thread a bit here: but the matrix is no longer stored permuted and is instead dynamically permuted as a view, correct?

File: src/linearmixedmodel.jl
Comment (palday) [inline src/linearmixedmodel.jl:974]: `k+1` because `k` is now just the number of `reterms` and we need to add one for the `feterm`, right?

File: test/matrixterm.jl
Comment (palday) [inline test/matrixterm.jl:28]: should we re-enable these tests?

File: test/matrixterm.jl
Comment (palday) [inline test/matrixterm.jl:49]:
```suggestion
    @testset "rank defiency in sparse FeTerm" begin
```

(contingent on any name changing)

File: test/matrixterm.jl
Comment (palday) [inline test/matrixterm.jl:71]:
```suggestion
end

```

(the diff highlighted the missing terminal newline)

File: test/pls.jl
Comment (palday) [inline test/pls.jl:34]: If you want to remove the `describeblocks` deprecation machinery as well, go for it! This will be the second major release after that deprecation.

File: src/linalg.jl
Comment (palday) [reply src/linalg.jl:34]: I think readability is about the same either way, so I can't be bothered to change it back. :smile:

**Comment by palday** [comment]
The failure on `Future` is something I've fixed as part of my changes in #474

File: src/Xymat.jl
Comment (dmbates) [reply src/Xymat.jl:6]: I'm a bit of a purist about the terms `design matrix` and `model matrix`.  To me the term `design matrix` refers to the experimental design, which is to say the levels of the factors in the data table.  The matrix derived from the data table and the formula is the model matrix.  So I would change that to `model matrix`.

Comment (palday) [reply src/Xymat.jl:6]: Easy! I actually originally wrote "model matrix" then remembered that the last time I wrote that, you wanted "design matrix". But now I understand the distinction you're drawing here.

File: src/linalg/logdet.jl
Comment (dmbates) [reply src/linalg/logdet.jl:36]: I don't think it should be too bad here because it is the log of the square root of the penalized residual sum of squares.  It shouldn't be especially large or small.  Also, this block is only used for REML estimation and the default is ML.

An alternative would be to define a function like `LD` that iterates over `diagind(size(d, 1), size(d, 2) - 1)` instead of `diagind(d)`.

File: src/Xymat.jl
Comment (dmbates) [reply src/Xymat.jl:75]: I have this feeling that it is easier to check the length of a vector than to get the number of columns in a matrix, although it probably involves a few nanoseconds at most.

File: src/Xymat.jl
Comment (dmbates) [reply src/Xymat.jl:104]: Okay, makes sense.

File: src/Xymat.jl
Comment (dmbates) [reply src/Xymat.jl:130]: Perhaps.  I'm also in the "no strong preference" camp.

Comment (palday) [reply src/Xymat.jl:130]: Then let's leave it as is because that's less effort.

**Comment by dmbates** [comment]
In `v"1.7.0-DEV.620"` and later, which use libblastrampoline, the first use of `BLAS.vendor()` produces a deprecation warning.  I can commit a change for that if it seems worthwhile.

File: test/pls.jl
Comment (dmbates) [reply test/pls.jl:34]: Okay, let's nuke `describeblocks` then.  Do you want to do it or shall I?

**Comment by palday** [comment]
Go for it -- we've already got a `@static` guard there noting that I would still like to have tests notify us of platform specifics (since those have been the source of problems in the past).

File: test/pls.jl
Comment (palday) [reply test/pls.jl:34]: I'll let you do it.

File: src/linearmixedmodel.jl
Comment (dmbates) [reply src/linearmixedmodel.jl:965]: Yes, I believe you are correct.  Good that you caught that.

File: test/pls.jl
Comment (dmbates) [reply test/pls.jl]: Done in https://github.com/JuliaStats/MixedModels.jl/pull/464/commits/3ee812d63fa74da7a9fe8f3b3150b2794801d4c7

**Comment by dmbates** [comment]
I think we should look at pivotedQR versus pivotedCholesky for determining the rank of X after we merge this.  So probably don't make a new release immediately.

**Comment by palday** [comment]
I also think there's a few things to think about before we put a bow on 4.0.

**Comment by palday** [comment]
There's also something I'm currently tracking down with the new `show` methods. In my local build of the docs, the markdown is used and renders correctly (not what we want for showing users how to step through the REPL, but also at least understandable). But the GHA builds all have the markdown output but not rendered into HTML. I'm working on figuring that out.

File: src/linalg/logdet.jl
Comment (palday) [reply src/linalg/logdet.jl:36]: I think this is fine for now, but if we ever run into weird trouble on REML fits, it'll be the first place I look.

**Comment by dmbates** [comment]
@palday Let me know if I should be looking at this PR or if you are ready to do the heavy git lifting to make it compatible with current master.

**Comment by palday** [comment]
I already did the lifting. :smile:

**Comment by dmbates** [comment]
I think this is ready to go.  If you agree @palday please go ahead and merge.

**Review by palday — APPROVED** [review]

**Pull request merged**
